File: Perl/IPC_HW/hw7.pl
#!/usr/bin/perl

our @stack;

sub show {
    return if ($_[0] == 0);
    show ($_[0] - 1);
    print "$stack[$_[0]-1] ";
}

sub qqq {
    my ($n, $k) = @_;
    if ($n < 0 || $k == 0) {
        return 0;
    }
    elsif ($n == 0) {
        show($#stack + 1);
        print "\n";
        return 1;
    }
    else {
        push @stack, $k;
        my $a = qqq($n-$k, $k);
        pop @stack;
        return $a + qqq($n, $k-1);
    }
}

sub ui {
    print "Enter a positive integer <= 20: ";
    if (defined ($_ = <>)) {
        chomp;
        printf ("There are %d ways to partition %d.\n", qqq($_, $_), $_);
        ui();
    }
    return;
}

ui();

File: Python/IPC_HW/hw7.py
#!/usr/bin/env python

import sys

stack = []

def show (n) :
    if n == 0 :
        return
    show (n - 1)
    sys.stdout.write ("%d " % stack[n - 1])


def q (n, k) :
    if n < 0 or k == 0:
        return 0
    elif n == 0 :
        show (len (stack))
        sys.stdout.write ("\n")
        return 1
    else :
        stack.append(k)
        a = q (n-k, k)
        stack.pop()
        return a + q (n, k-1)


def ui ():
    try :
        n = int (raw_input ("Enter a positive integer <= 20: "))
        sys.stdout.write ("There are %d ways to partition %d.\n" % (q (n, n), n))
        ui()
    except EOFError :
        return


if __name__ == '__main__' :
    ui() 
